dask/bag/avro.py

```diff
@@ -0,0 +1,126 @@
+import copy
+import io
+
+MAGIC = b'Obj\x01'
+SYNC_SIZE = 16
+
+
+def read_long(fo):
+    """variable-length, zig-zag encoding."""
+    c = fo.read(1)
+    b = ord(c)
+    n = b & 0x7F
+    shift = 7
+    while (b & 0x80) != 0:
+        b = ord(fo.read(1))
+        n |= (b & 0x7F) << shift
+        shift += 7
+    return (n >> 1) ^ -(n & 1)
+
+
+def read_bytes(fo):
+    """a long followed by that many bytes of data."""
+    size = read_long(fo)
+    return fo.read(size)
+
+
+def read_header(fo):
+    """Extract an avro file's header
+
+    fo: file-like
+        This should be in bytes mode, e.g., io.BytesIO
+
+    Returns dict representing the header
+
+    Parameters
+    ----------
+    fo: file-like
+    """
+    assert fo.read(len(MAGIC)) == MAGIC, 'Magic avro bytes missing'
+    meta = {}
+    out = {'meta': meta}
+    while True:
+        n_keys = read_long(fo)
+        if n_keys == 0:
+            break
+        for i in range(n_keys):
+            # ignore dtype mapping for bag version
+            read_bytes(fo)  # schema keys
+            read_bytes(fo)  # schema values
+    out['sync'] = fo.read(SYNC_SIZE)
+    out['header_size'] = fo.tell()
+    fo.seek(0)
+    out['head_bytes'] = fo.read(out['header_size'])
+    return out
+
+
+def read_avro(urlpath, blocksize=100000000, storage_options=None):
+    """Read set of avro files
+
+    Use this with arbitrary nested avro schemas. Please refer to the
+    fastavro documentation for its capabilities:
+    https://github.com/fastavro/fastavro
+
+    Parameters
+    ----------
+    urlpath: string or list
+        Absolute or relative filepath, URL (may include protocols like
+        ``s3://``), or globstring pointing to data.
+    blocksize: int or None
+        Size of chunks in bytes. If None, there will be no chunking and each
+        file will become one partition.
+    storage_options: dict or None
+        passed to backend file-system
+    """
+    from dask.utils import import_required
+    from dask import delayed
+    from dask.bytes.core import open_files, read_bytes
+    from dask.bag import from_delayed
+    import_required('fastavro',
+                    "fastavro is a required dependency for using "
+                    "bag.read_avro().")
+
+    storage_options = storage_options or {}
+    files = open_files(urlpath, **storage_options)
+    with copy.copy(files[0]) as f:
+        # we assume the same header for all files
+        head = read_header(f)
```

Comment:
Please forgive the naive question, but what exactly does this header represent?

Mainly wondering if it could change if the avro files in a given urlpath had different schemas..

Reply:
The header has a name, description, namespace, schema (fields and types) and a block sync marker. You are right, in the case of bags, it may be OK to read files where each has a different schema, and I actually think the comment is wrong and this would work fine as things stand. There is no test for different schemas, though, I hadn't considered it.

Reply:
Ah, actually I've made a mistake - we need the sync marker before up front, else the read_bytes call above goes wrong. This'll take a little bit of work.

Reply:
I see what you are saying, bit of a circular dependency. In the `dask-avro` library @rmax used the [fastavro reader to grab the sync marker before calling `read_bytes`](https://github.com/rmax/dask-avro/blob/master/src/dask_avro/__init__.py#L59-L60) if that helps.

Either way, if we can get by with just reading the header from one file that would be great, as I think doing it for each file significantly slowed things down in dask_avro.

Reply:
(see my comment below)
In general, the sync markers are all different. I think it is possibly to have identical headers for every file, but that is not the normal way to do things.